iscservice/demo/libdemo.c

```diff
@@ -0,0 +1,352 @@
+			     uint32_t shift)
+{
+	int   rc;
+	int   i;
+	char *p;
+
+	p = m0_alloc_aligned(m0_io_count(iiv), shift);
+	if (p == NULL)
+		return M0_ERR_INFO(-ENOMEM, "failed to allocate buf");
+
+	rc = m0_bufvec_empty_alloc(bv, iiv->ci_nr);
+	if (rc != 0) {
+		m0_free(p);
+		return M0_ERR_INFO(rc, "failed to allocate bufvec");
+	}
+
+	for (i = 0; i < iiv->ci_nr; i++) {
+		bv->ov_buf[i] = p;
+		bv->ov_vec.v_count[i] = iiv->ci_iosegs[i].ci_count;
+		p += iiv->ci_iosegs[i].ci_count;
+	}
+
+	return 0;
+}
+
+int launch_stob_io(struct m0_isc_comp_private *pdata,
+		   struct m0_buf *in, int *rc)
+{
+	uint32_t           shift = 0;
+	struct m0_stob_io *stio = (struct m0_stob_io *)pdata->icp_data;
+	struct m0_fom     *fom = pdata->icp_fom;
+	struct isc_targs   ta = {};
+	struct m0_stob_id  stob_id;
+	struct m0_stob    *stob = NULL;
+
+	*rc = m0_xcode_obj_dec_from_buf(&M0_XCODE_OBJ(isc_targs_xc, &ta),
+					in->b_addr, in->b_nob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to xdecode args: rc=%d", *rc);
+		return M0_FSO_AGAIN;
+	}
+
+	if (ta.ist_ioiv.ci_nr == 0) {
+		M0_LOG(M0_ERROR, "no io segments given");
+		*rc = -EINVAL;
+		goto err;
+	}
+
+	m0_stob_io_init(stio);
+	stio->si_opcode = SIO_READ;
+
+	m0_fid_convert_cob2stob(&ta.ist_cob, &stob_id);
+	*rc = m0_storage_dev_stob_find(m0_cs_storage_devs_get(),
+				       &stob_id, &stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to find stob by cob="FID_F": rc=%d",
+		       FID_P(&ta.ist_cob), *rc);
+		goto err;
+	}
+
+	shift = m0_stob_block_shift(stob);
+
+	*rc = m0_indexvec_wire2mem(&ta.ist_ioiv, ta.ist_ioiv.ci_nr,
+				   shift, &stio->si_stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to make cob ivec: rc=%d", *rc);
+		goto err;
+	}
+
+	*rc = bufvec_alloc_init(&stio->si_user, &ta.ist_ioiv, shift);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to allocate bufvec: rc=%d", *rc);
+		goto err;
+	}
+
+	bufvec_pack(&stio->si_user, shift);
+
+	*rc = m0_stob_io_private_setup(stio, stob);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to setup adio for stob="FID_F": rc=%d",
+		       FID_P(&stob->so_id.si_fid), *rc);
+		goto err;
+	}
+
+	/* make sure the fom is waked up on I/O completion */
+	m0_mutex_lock(&stio->si_mutex);
+	m0_fom_wait_on(fom, &stio->si_wait, &fom->fo_cb);
+	m0_mutex_unlock(&stio->si_mutex);
+
+	*rc = m0_stob_io_prepare_and_launch(stio, stob, NULL, NULL);
+	if (*rc != 0) {
+		M0_LOG(M0_ERROR, "failed to launch io for stob="FID_F": rc=%d",
+		       FID_P(&stob->so_id.si_fid), *rc);
+		m0_mutex_lock(&stio->si_mutex);
+		m0_fom_callback_cancel(&fom->fo_cb);
+		m0_mutex_unlock(&stio->si_mutex);
+	}
+ err:
+	if (*rc != 0) {
+		if (stob != NULL) {
+			bufvec_open(&stio->si_user, shift);
+			stio_fini(stio, stob);
+		}
+		/*
+		 * EAGAIN has a special meaning in the calling isc code,
+		 * so make sure we don't return it by accident.
+		 */
+		if (*rc == -EAGAIN)
+			*rc = -EBUSY;
+		return M0_FSO_AGAIN;
+	}
+
+	*rc = -EAGAIN; /* Wait for the I/O result. */
+	return M0_FSO_WAIT;
+}
```

Comment:
Can we create a function, e.g. `stob_read(fid, off, len, buf)`?
And then hide all the details of how to do it.
Developers should only know that we are going to read some data from a stob, and compute MIN/MAX from the data (treat the data as array of float number).

Reply:
I'm afraid, the current ISC API does not allow to do this. The computation function is called two times: 1) fetch the data from stob (in asynchronous way); 2) do the computation on the fetched data.

Even if we introduce such asynchronous `stob_read()` function into Motr API, we need to provide to it also at least `stio` and `fom` (so that it could be woken up when the I/O is completed). In addition, after the I/O is completed, we need to call `bufvec_open()` to shift the result to normal values.

Will it make the code simpler in the end? I'm not sure. Anyway, let me give it a try...

Reply:
Introduced `m0_isc_io_launch()` into libmotr, please check.